core/src/cubos/core/al/oal_audio_device.cpp

```diff
@@ -0,0 +1,202 @@
+#include <cubos/core/al/oal_audio_device.hpp>
+#include <cubos/core/log.hpp>
+
+#include <AL/al.h>
+#include <AL/alc.h>
+
+using namespace cubos::core::al;
+
+class OALBuffer : public impl::Buffer
+{
+public:
+    OALBuffer(ALuint id) : id(id)
+    {
+    }
+
+    virtual ~OALBuffer() override
+    {
+        alDeleteBuffers(1, &this->id);
+    }
+
+    virtual void fill(Format format, size_t size, const void* data, size_t frequency) override
+    {
+        ALenum alFormat = 0;
+
+        switch (format)
+        {
+        case Format::Mono8:
+            alFormat = AL_FORMAT_MONO8;
+            break;
+
+        case Format::Mono16:
+            alFormat = AL_FORMAT_MONO16;
+            break;
+
+        case Format::Stereo8:
+            alFormat = AL_FORMAT_STEREO8;
+            break;
+
+        case Format::Stereo16:
+            alFormat = AL_FORMAT_STEREO16;
+            break;
+        }
+
+        alBufferData(this->id, alFormat, data, size, frequency);
+    }
+
+    ALuint id;
+};
+
+class OALSource : public impl::Source
+{
+public:
+    OALSource(ALuint id) : id(id)
+    {
+    }
+
+    virtual ~OALSource() override
+    {
+        alDeleteSources(1, &this->id);
+    }
+
+    virtual void setBuffer(Buffer buffer) override
+    {
+        auto oalBuffer = std::dynamic_pointer_cast<OALBuffer>(buffer);
+        alSourcei(this->id, AL_BUFFER, oalBuffer->id);
```

Comment:
Why is this needed? Can't the id value be supplied directly?
```suggestion
        alSourcei(this->id, AL_BUFFER, buffer.id);
```